src/server/snapshot.cc

```diff
@@ -95,251 +94,243 @@ void SliceSnapshot::Join() {
         this_fiber::yield();
-        last_yield = serialized_;
         DVLOG(2) << "After sleep";
 
+        last_yield = stats_.serialized;
         // flush in case other fibers (writes commands that pushed previous values)
         // filled the buffer.
-        FlushSfile(false);
+        FlushDefaultBuffer(false);
       }
     } while (cursor);
 
     DVLOG(2) << "after loop " << this_fiber::properties<FiberProps>().name();
-    FlushSfile(true);
+    FlushDefaultBuffer(true);
   }  // for (dbindex)
 
   // Wait for SerializePhysicalBucket to finish.
   mu_.lock();
   mu_.unlock();
 
   for (unsigned i = 10; i > 1; i--)
-    CHECK(!rdb_serializer_->SendFullSyncCut());
-  FlushSfile(true);
+    CHECK(!default_serializer_->SendFullSyncCut());
+  FlushDefaultBuffer(true);
 
-  VLOG(1) << "Exit SnapshotSerializer (serialized/side_saved/cbcalls): " << serialized_ << "/"
-          << side_saved_ << "/" << savecb_calls_;
+  VLOG(1) << "Exit SnapshotSerializer (serialized/side_saved/cbcalls): " << stats_.serialized << "/"
+          << stats_.side_saved << "/" << stats_.savecb_calls;
 }
 
-void SliceSnapshot::CloseRecordChannel() {
-  // stupid barrier to make sure that SerializePhysicalBucket finished.
-  // Can not think of anything more elegant.
-  mu_.lock();
-  mu_.unlock();
+bool SliceSnapshot::BucketSaveCb(PrimeIterator it) {
+  // if we touched that physical bucket - skip it.
+  // We must make sure we TraverseBucket exactly once for each physical bucket.
+  // This test is the first one because it's likely to be the fastest one:
+  // physical_mask_ is likely to be loaded in L1 and bucket_id() does not require accesing the
+  // prime_table.
+  /*if (physical_mask_.test(it.bucket_id())) {
+    return false;
+  }*/
 
-  // Make sure we close the channel only once with a CAS check.
-  bool actual = false;
-  if (closed_chan_.compare_exchange_strong(actual, true)) {
-    dest_->StartClosing();
+  ++stats_.savecb_calls;
+
+  uint64_t v = it.GetVersion();
+  if (v >= snapshot_version_) {
+    // either has been already serialized or added after snapshotting started.
+    DVLOG(3) << "Skipped " << it.segment_id() << ":" << it.bucket_id() << ":" << it.slot_id()
+             << " at " << v;
+    ++stats_.skipped;
+    return false;
   }
+
+  ++stats_.serialized += SerializeBucket(current_db_, it);
+  return false;
+}
+
+unsigned SliceSnapshot::SerializeBucket(DbIndex db_index, PrimeTable::bucket_iterator it) {
+  DCHECK_LT(it.GetVersion(), snapshot_version_);
+
+  // traverse physical bucket and write it into string file.
+  it.SetVersion(snapshot_version_);
+  unsigned result = 0;
+
+  lock_guard lk(mu_);
+
+  BorrowedSerializer borrowed{this, db_index, compression_mode_ != CompressionMode::NONE};
+  RdbSerializer* serializer = borrowed.Get();
+
+  while (!it.is_done()) {
+    ++result;
+    SerializeEntry(db_index, it->first, it->second, nullopt, serializer);
+    ++it;
+  }
+  borrowed.num_records += result;
```

Comment:
And it allows using a RdbSerializer transparently without duplicating a lot of code and making the function unnecessary big and complex